synapse/handlers/stats.py

```diff
@@ -52,4 +59,324 @@ def __init__(self, hs):
     def notify_new_event(self):
         """Called when there may be more deltas to process
         """
-        pass
+        if not self.hs.config.stats_enabled:
+            return
+
+        lock = self.store.stats_delta_processing_lock
```

Comment:
We generally just use a boolean flag rather than a lock, it makes it less likely that someone accidentally does `lock.acquire()` and blocks the thread

Reply:
I kind of need a lock because the stats regenerator will want to wait on it.

However, what isn't clear to me now is whether the stats regenerator (a background process) will actually be ran in a different thread. That's the impression I got at the time but now I'm having trouble confirming it. Do you know @erikjohnston ?

Reply:
You basically never want to block. Either you're on the main thread where you'd block the reactor or you're on a database thread where you'd take out a transaction. 

There's also https://twistedmatrix.com/documents/current/api/twisted.internet.defer.DeferredLock.html, which may be what you actually want.